[quant][graphmode] linear_relu

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36994 [quant][graphmode] Test refactor -- extract graph checker
* **#36889 [quant][graphmode] linear_relu**

Differential Revision: [D21111540](https://our.internmc.facebook.com/intern/diff/D21111540)

Commit:
[quant][graphmode] linear_relu

[ghstack-poisoned]

Commit:
Update on "[quant][graphmode] linear_relu"


Differential Revision: [D21111540](https://our.internmc.facebook.com/intern/diff/D21111540)

[ghstack-poisoned]

z-a-f commented on torch/quantization/qconfig.py:
```diff
@@ -18,7 +18,7 @@ class QConfig(namedtuple('QConfig', ['activation', 'weight'])):
     method (that behaves like functools.partial):
 
-      my_qconfig = QConfig(activation=MinMaxObserver.with_args(dtype=torch.qint8), 
+      my_qconfig = QConfig(activation=MinMaxObserver.with_args(dtype=torch.qint8),
```
Sorry, this one happened because my editor removes all the end lines, and this file was open :/

Commit:
Update on "[quant][graphmode] linear_relu"


Differential Revision: [D21111540](https://our.internmc.facebook.com/intern/diff/D21111540)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1169,12 +1169,58 @@ def _test_op_impl(self, module, data, quantized_op):
 
         return model
 
+    @unittest.skipUnless(
```
Do we need to skip QNNPACK?

jerryzh168 replied:
not sure if qnnpack works, my backward compatibility PR added qnnpack and the test failed

z-a-f replied:
When I enable the qnnpack locally, the tests pass

z-a-f replied:
I take it back -- the tests fail. Specifically line 950 in `observer.py` throws `RuntimeError: RuntimeError: Trying to create tensor with negative dimension -4398046511104`. I think that might be related to #36888

raghuramank100 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1169,12 +1169,58 @@ def _test_op_impl(self, module, data, quantized_op):
+            def __init__(self):
+                super(M, self).__init__()
+                self.linear = torch.nn.Linear(30, 4).float()
+                self.relu = torch.nn.ReLU()
```
Do we need to test functional relu as well?
For completeness, should we also test functional linear + functional relu?

Commit:
Update on "[quant][graphmode] linear_relu"


Differential Revision: [D21111540](https://our.internmc.facebook.com/intern/diff/D21111540)

[ghstack-poisoned]

Commit:
Update on "[quant][graphmode] linear_relu"


Differential Revision: [D21111540](https://our.internmc.facebook.com/intern/diff/D21111540)

[ghstack-poisoned]

Commit:
Update on "[quant][graphmode] linear_relu"


Differential Revision: [D21111540](https://our.internmc.facebook.com/intern/diff/D21111540)

[ghstack-poisoned]

Commit:
Update on "[quant][graphmode] linear_relu"


Differential Revision: [D21111540](https://our.internmc.facebook.com/intern/diff/D21111540)

[ghstack-poisoned]

Commit:
Update on "[quant][graphmode] linear_relu"


Differential Revision: [D21111540](https://our.internmc.facebook.com/intern/diff/D21111540)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -667,6 +667,20 @@ class InsertObserversHelper {
   // These are the IR patterns we match to skip inserting observers.
   // They are compiled once on construction and used repeatedly within
   // the pass.
+  const PatternInfo linear_functional_relu = PatternInfo::parse_from_str(R"(
```
maybe add the fusion for F.linear + F.relu and F.linear + self.relu(module) here?
We may need these two patterns for F.conv as well, but it can be in a different PR.

z-a-f replied:
OK, I will look into this

Commit:
Update on "[quant][graphmode] linear_relu"


Differential Revision: [D21111540](https://our.internmc.facebook.com/intern/diff/D21111540)

[ghstack-poisoned]

Commit:
Update on "[quant][graphmode] linear_relu"


Differential Revision: [D21111540](https://our.internmc.facebook.com/intern/diff/D21111540)

[ghstack-poisoned]

Commit:
Update on "[quant][graphmode] linear_relu"


Differential Revision: [D21111540](https://our.internmc.facebook.com/intern/diff/D21111540)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1168,13 +1168,126 @@ def _test_op_impl(self, module, data, quantized_op):
+                self.b = torch.randn(4)
+
+            def forward(self, x):
+                return self.linear(x, self.w, self.b)
```
you can just call F.linear here

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1168,13 +1168,126 @@ def _test_op_impl(self, module, data, quantized_op):
+        " Quantized operations require FBGEMM. FBGEMM is only optimized for CPUs"
+        " with instruction set support avx2 or newer.",
+    )
+    def test_quantized_functional_linear_relu(self):
```
actually, can you merge this test with the previous one?

z-a-f replied:
I am not sure I can -- I tried doing it, but that creates if statements in the forward, which fails to match the patterns :'(

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -766,6 +801,10 @@ graph(%self, %a, %b, %inplace):
 
   const std::vector<std::reference_wrapper<const PatternInfo>> delay_patterns =
       {
+          nn_linear_nn_relu,
+          nn_linear_f_relu,
+          f_linear_nn_relu,
+          f_linear_f_relu,
```
this looks like a great naming convention, we should probably follow the same for other patterns in the list.

z-a-f replied:
Will do when adding it to conv

jerryzh168 reviewed (approved):
It's fine to keep the tests separate if it's hard to merge them

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1168,13 +1168,126 @@ def _test_op_impl(self, module, data, quantized_op):
+
+        data = [(torch.randn((1, 30), dtype=torch.float),
+                 torch.randint(0, 1, (1,), dtype=torch.long)) for _ in range(2)]
+        model = self._test_op_impl(M(functional=False), data,
+                                   "quantized::linear_relu")
+        model_functional = self._test_op_impl(M(functional=True), data,
+                                              "quantized::linear_relu")
+        checker = FileCheck().check_not("aten::linear") \
+                             .check_not("aten::relu") \
+                             .check_not("quantized::linear(") \
+                             .check_not("quantized::relu(")
+        checker.run(model.graph)
+        checker.run(model_functional.graph)
```
probably better to loop over True and False here and put the call to _test_op_impl and checker.run in the loop

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1168,13 +1168,126 @@ def _test_op_impl(self, module, data, quantized_op):
+
+        data = [(torch.randn((1, 30), dtype=torch.float),
+                 torch.randint(0, 1, (1,), dtype=torch.long)) for _ in range(2)]
+        model = self._test_op_impl(M(functional=False), data,
+                                   "quantized::linear_relu")
+        model_functional = self._test_op_impl(M(functional=True), data,
+                                              "quantized::linear_relu")
+        checker = FileCheck().check_not("aten::linear") \
+                             .check_not("aten::relu") \
+                             .check_not("quantized::linear(") \
+                             .check_not("quantized::relu(")
+        checker.run(model.graph)
+        checker.run(model_functional.graph)
```
same here

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -665,9 +665,44 @@ class InsertObserversHelper {
   // the pass.
+
+  // nn.Linear + nn.ReLU
+  const PatternInfo nn_linear_nn_relu = PatternInfo::parse_from_str(R"(
```
btw these patterns will need to be updated, please see the code in master to see how to do the update

z-a-f commented:
Replaced by #40021
